File: frontend/src/lib/api/behaviour.ts
import { api } from './client';

export type Severity = 'LOW' | 'MEDIUM' | 'HIGH';

export const SEVERITY_LABELS: Record<Severity, string> = {
  LOW: 'Low', MEDIUM: 'Medium', HIGH: 'High',
};

export interface BehaviourRecord {
  id: string;
  student_id: string;
  academic_term_id: string;
  incident_type: string;
  description: string;
  severity: Severity;
  action_taken: string | null;
  incident_date: string;
  recorded_by_id: string;
  created_at: string;
}

export const listBehaviourRecords = (studentId: string, termId: string): Promise<BehaviourRecord[]> =>
  api.get('/behaviour', { params: { student_id: studentId, term_id: termId } }).then(r => r.data);

export const createBehaviourRecord = (data: {
  student_id: string;
  academic_term_id: string;
  incident_type: string;
  description: string;
  severity: Severity;
  action_taken?: string;
  incident_date: string;
  override_reason?: string;
}): Promise<BehaviourRecord> =>
  api.post('/behaviour', data).then(r => r.data);

export const deleteBehaviourRecord = (recordId: string, overrideReason?: string): Promise<void> =>
  api.delete(`/behaviour/${recordId}`, {
    params: overrideReason ? { override_reason: overrideReason } : undefined,
  }).then(() => undefined);

File: backend/app/schemas/behaviour.py
from __future__ import annotations
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, field_validator


class BehaviourRecordCreate(BaseModel):
    student_id: uuid.UUID
    academic_term_id: uuid.UUID
    incident_type: str
    description: str
    severity: Literal["LOW", "MEDIUM", "HIGH"]
    action_taken: str | None = None
    incident_date: date
    # Only required when the term has results_locked=True.
    override_reason: str | None = None

    @field_validator("incident_type", "description")
    @classmethod
    def not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Field must not be blank")
        return v.strip()


class BehaviourRecordRead(BaseModel):
    id: uuid.UUID
    student_id: uuid.UUID
    academic_term_id: uuid.UUID
    incident_type: str
    description: str
    severity: str
    action_taken: str | None
    incident_date: date
    recorded_by_id: uuid.UUID
    created_at: datetime
    model_config = {"from_attributes": True}


class BehaviourAuditLogRead(BaseModel):
    id: uuid.UUID
    behaviour_record_id: uuid.UUID | None
    student_id: uuid.UUID
    action: str
    incident_type: str
    incident_date: date
    changed_by_id: uuid.UUID
    reason: str | None
    changed_at: datetime
    model_config = {"from_attributes": True}
